src/execution/execute-afm.ts

```diff
@@ -1,137 +1,251 @@
-        for (let i = 0; i < data.length; i += 1) {
-            const columns = data[i] as any;
+function validateNumOfDimensions(numOfDimensions: number): void {
+    invariant(
+        numOfDimensions === 1 || numOfDimensions === 2,
+        `${numOfDimensions} dimensions are not allowed. Only 1 or 2 dimensions are supported.`
+    );
+}
 
-            result.executionResult.data[i + rowOffset].push(...columns);
-        }
-    } else { // appending new rows
-        result.executionResult.data.push(...data);
+function replaceLimitAndOffsetInUri(oldUri: string, limit: number[], offset: number[]): string {
+    const [uriPart, queryPart] = oldUri.split(/\?(.+)/);
+    const query = {
+        ...qs.parse(queryPart),
+        limit: limit.join(','),
+        offset: offset.join(',')
+    };
+
+    return uriPart + qs.stringify(query, { addQueryPrefix: true });
+}
+
+export function getNextOffset(limit: number[], offset: number[], total: number[]): number[] {
+    const numOfDimensions = total.length;
+    validateNumOfDimensions(numOfDimensions);
+
+    const defaultNextRowsOffset = offset[0] + limit[0];
+
+    if (numOfDimensions === 1) {
+        return [defaultNextRowsOffset];
+    }
+
+    const defaultNextColumnsOffset = offset[1] + limit[1];
+
+    const nextColumnsExist = offset[1] + limit[1] < total[1];
+
+    const nextRowsOffset = nextColumnsExist
+        ? offset[0] // stay in the same rows
+        : defaultNextRowsOffset; // go to the next rows
+
+    const nextColumnsOffset = nextColumnsExist
+        ? defaultNextColumnsOffset // next columns for the same rows
+        : 0; // start in the beginning of the next rows
+
+    return [nextRowsOffset, nextColumnsOffset];
+}
+
+export function getNextLimit(limit: number[], nextOffset: number[], total: number[]): number[] {
+    const numOfDimensions = total.length;
+    validateNumOfDimensions(numOfDimensions);
+
+    const getSingleNextLimit = (limit: number, nextOffset: number, total: number): number => nextOffset + limit > total
+        ? total - nextOffset
+        : limit;
+
+    // prevent set up lower limit than possible for 2nd dimension in the beginning of the next rows
+    if (
+        numOfDimensions === 2 &&
+        nextOffset[1] === 0 && // beginning of the next rows
+        limit[0] < total[1] // limit from 1st dimension should be used in 2nd dimension
+    ) {
+        return [
+            getSingleNextLimit(limit[0], nextOffset[0], total[0]),
+            limit[0]
+        ];
     }
 
-    const doDimension = (dim: number) => {
-        const otherDimension = (dim === 0 ? 1 : 0);
+    return range(numOfDimensions).map((i: number) => getSingleNextLimit(limit[i], nextOffset[i], total[i]));
+}
+
+export function nextPageExists(nextOffset: number[], total: number[]): boolean {
+    // expression "return nextLimit[0] > 0" also returns correct result
+    return nextOffset[0] < total[0];
+}
+
+// works only for one or two dimensions
+export function mergePage(
```

Comment:
this looks to me that mergePage function is solving more than one thing

at least it looks to me it could be split into separate methods 

// merge data
// merge headerItems
// update page count

what do you think?